pgx-pg-sys/src/lib.rs

```diff
@@ -269,6 +273,66 @@ impl AsPgCStr for &Option<std::string::String> {
     }
 }
 
+pub trait PgNodeCast<T> {
+    fn cast_from(pg_node: T) -> Self
+    where
+        Self: Sized;
+}
+
+pub trait PgNodeTryCast<T> {
+    fn try_cast_from(pg_node: T) -> Option<Self>
+    where
+        Self: Sized;
+}
+
+impl<A: PgNode, B: PgNode> PgNodeCast<&A> for &B {
+    fn cast_from(pg_node: &A) -> Self {
+        Self::try_cast_from(pg_node).unwrap()
+    }
+}
+
+impl<A: PgNode, B: PgNode> PgNodeTryCast<&A> for &B {
+    fn try_cast_from(pg_node: &A) -> Option<Self> {
+        unsafe {
+            let node = std::mem::transmute::<_, &Node>(pg_node);
+            match B::NODE_TAG {
+                Some(tag) => {
+                    if node.type_ == tag {
+                        Some(std::mem::transmute::<_, &B>(node))
+                    } else {
+                        None
+                    }
+                }
+                None => None,
+            }
+        }
```

Comment:
Hrmmm. What "directionality" of casting is this intended to handle? Upcasting? Downcasting? Something else? I feel we should support checked casting, for sure, instead of just making people do pointer-casts, but I don't feel confident about what this is "for", basically.